table/table.go

```diff
@@ -219,61 +220,49 @@ func (t *Table) readNoFail(off, sz int) []byte {
 }
 
 func (t *Table) readIndex() error {
-	if len(t.mmap) != t.tableSize {
-		panic("Table size does not match the read bytes")
-	}
 	readPos := t.tableSize
 
-	// Read bloom filter.
+	// Read checksum len from the last 4 bytes
 	readPos -= 4
 	buf := t.readNoFail(readPos, 4)
-	bloomLen := int(binary.BigEndian.Uint32(buf))
-	readPos -= bloomLen
-	data := t.readNoFail(readPos, bloomLen)
-	t.bf = bbloom.JSONUnmarshal(data)
-
-	readPos -= 4
-	buf = t.readNoFail(readPos, 4)
-	restartsLen := int(binary.BigEndian.Uint32(buf))
-
-	readPos -= 4 * restartsLen
-	buf = t.readNoFail(readPos, 4*restartsLen)
+	checksumLen := binary.BigEndian.Uint32(buf)
 
-	offsets := make([]int, restartsLen)
-	for i := 0; i < restartsLen; i++ {
-		offsets[i] = int(binary.BigEndian.Uint32(buf[:4]))
-		buf = buf[4:]
+	// Read checksum
+	checksumExpected := pb.Checksum{}
+	readPos -= int(checksumLen)
+	buf = t.readNoFail(readPos, int(checksumLen))
+	if err := checksumExpected.Unmarshal(buf); err != nil {
+		return err
 	}
 
-	// The last offset stores the end of the last block.
-	for i := 0; i < len(offsets); i++ {
-		var o int
-		if i == 0 {
-			o = 0
-		} else {
-			o = offsets[i-1]
+	// Read index size from the footer
+	readPos -= 4
+	buf = t.readNoFail(readPos, 4)
+	indexLen := int(binary.BigEndian.Uint32(buf))
+	// Read index
+	readPos -= indexLen
+	data := t.readNoFail(readPos, indexLen)
+
+	if checksumExpected.Algo == pb.Checksum_CRC32C {
+		checksum := crc32.Checksum(data, y.CastagnoliCrcTable)
+		if checksum != checksumExpected.Sum32 {
```

Comment:
can move this outside the outer if

Reply:
@mangalaman93 It wouldn't really simplify code.
```go
	if checksumExpected.Algo == pb.Checksum_CRC32C {
		checksum := crc32.Checksum(data, y.CastagnoliCrcTable)
		if checksum != checksumExpected.Sum32 {
			return y.Wrapf(ErrChecksumMismatch, "actual: %d, expected: %d", checksum, checksumExpected)
		}
	} else if checksumExpected.Algo == pb.Checksum_XXHash {
		checksum := xxhash.Sum64(data)
		if checksum != checksumExpected.Sum64 {
			return y.Wrapf(ErrChecksumMismatch, "actual: %d, expected: %d", checksum, checksumExpected)
		}
	} else {
		return errors.New("Unknown checksum type")
	}
```
Note that `checkumExpected.Sum32` is being used at one place and `checkusmExpected.Sum64` is being used in another place.

The checksum struct has two fields `Sum32` and `Sum64`. The fields are populated based on the type of checksum used.